### PR TITLE
fix(models): pass mask and is_causal through gradient checkpointing in MaskedCausalVisionTransformer (#2049)

### LINKED ISSUE #2049: MaskedCausalVisionTransformer ignores masks with gradient checkpointing
`MaskedCausalVisionTransformer.forward_features()` ignores its attention mask when gradient checkpointing is enabled. The checkpointed masked output matches an unmasked forward instead of the ordinary masked path.

```python
import torch

from lightly.models.modules.masked_causal_vision_transformer import MaskedCausalVisionTransformer

torch.manual_seed(0)
model = MaskedCausalVisionTransformer(
    img_size=32, patch_size=16, embed_dim=24, depth=2, num_heads=3
).eval()
images = torch.rand(2, 3, 32, 32)
mask = torch.zeros(2, 5, dtype=torch.bool)
mask[:, 1:] = True

expected = model.forward_features(images, mask=mask)
model.set_grad_checkpointing()
actual = model.forward_features(images, mask=mask)

torch.testing.assert_close(actual, expected)
```

Environment: Linux 6.8, Python 3.12.3, PyTorch 2.13.0 CPU.

### PR DESCRIPTION
**Pull request by harmehak0173**
Fixes #2049

### Summary
When gradient checkpointing is enabled (`model.set_grad_checkpointing(True)`), `MaskedCausalVisionTransformer.forward_features()` previously used `_manipulate.checkpoint_seq(self.blocks, x)`, which only forwarded tensor `x` and dropped the `mask` and `is_causal` arguments. Consequently, attention masking was ignored when checkpointing was enabled.

### Changes
1. `lightly/models/modules/masked_causal_vision_transformer.py`:
   - Updated `MaskedCausalBlock.forward` to forward `is_causal` to `self.attn`.
   - Updated `MaskedCausalVisionTransformer.forward_features` to checkpoint each block with `torch.utils.checkpoint.checkpoint(block, x, mask, is_causal, use_reentrant=False)`, preserving the `mask` and `is_causal` arguments during both forward and backward passes.
   - Removed unused `_manipulate` import.
2. `tests/models/modules/test_masked_causal_vision_transformer.py`:
   - Added unit tests verifying output parity between checkpointed and non-checkpointed forward passes with masked and unmasked inputs.
   - Added unit tests verifying gradient parity (parameter gradients and input gradients) during backward pass with gradient checkpointing enabled.
   - Added tests for `is_causal=True` and `is_causal=False`.

- Preserved `mask` and `is_causal` during gradient-checkpointed execution.
- Added output and gradient parity tests for causal and non-causal modes.
- Removed the unused `_manipulate` import.

### CODE REVIEW

File: lightly/models/modules/masked_causal_vision_transformer.py [hunk 1]
```diff
@@ -151,7 +150,9 @@ def forward(
         Returns:
             Output tensor after applying the attention block.
         """
-        x = x + self.drop_path1(self.ls1(self.attn(self.norm1(x), mask=mask)))
+        x = x + self.drop_path1(
+            self.ls1(self.attn(self.norm1(x), mask=mask, is_causal=is_causal))
```
[the remark this answers is not in the document]
Comment (harmehak0173) [reply lightly/models/modules/masked_causal_vision_transformer.py:153]: Updated `MaskedCausalAttention.forward` and `_get_attention_mask` to pass `is_causal`, ensuring the causal mask is only constructed/applied when `is_causal=True`. Also added `test_forward_features__is_causal_controls_masked_attention` in `test_masked_causal_vision_transformer.py` to verify that `is_causal=False` enables bidirectional attention matching unmasked inputs.

**Comment by gabrielfruet** [comment]
@harmehak0173 Hi, could you fix the issues and merge conflicts?

**Comment by gabrielfruet** [comment]
Hi, are you still interested in working on this? If not, I’ll offer it to someone else.
